File: frontend/src/microservices/advertising/objects/CreateAdvertisementRequest.ts
export type CreateAdvertisementRequest = {
  ownerManagerId: string
  ownerType: string
  ownerDisplayName: string
  advertisementKind?: string | null
  title: string
  subtitle: string
  description: string
  imageUrl?: string | null
  ctaLabel: string
  targetResourceType: string
  targetResourceId: string
  resourceSummaryTitle?: string | null
  landingTarget?: string | null
  placement: string
  creativeJson?: string | null
  creativeWidth?: number | null
  creativeHeight?: number | null
  priority: number
  startAt: string
  endAt: string
}

export const createAdvertisementRequestFromJson = (json: string): CreateAdvertisementRequest =>
  JSON.parse(json) as CreateAdvertisementRequest

export const createAdvertisementRequestToJson = (value: CreateAdvertisementRequest): string =>
  JSON.stringify(value)

File: backend/src/main/scala/microservices/advertising/objects/CreateAdvertisementRequest.scala
// 本文件定义 advertising 模块的 CreateAdvertisementRequest，作为后端与前端同名的数据模型，并提供 JSON codec。
package com.typesafe.travel.advertising.domain

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

final case class CreateAdvertisementRequest(
    ownerManagerId: String,
    ownerType: String,
    ownerDisplayName: String,
    advertisementKind: Option[String],
    title: String,
    subtitle: String,
    description: String,
    imageUrl: Option[String],
    ctaLabel: String,
    targetResourceType: String,
    targetResourceId: String,
    resourceSummaryTitle: Option[String],
    landingTarget: Option[String],
    placement: String,
    creativeJson: Option[String],
    creativeWidth: Option[Int],
    creativeHeight: Option[Int],
    priority: Int,
    startAt: String,
    endAt: String
)

object CreateAdvertisementRequest:
  given sourceEncoder: Encoder[CreateAdvertisementRequest] = deriveEncoder
  given sourceDecoder: Decoder[CreateAdvertisementRequest] = deriveDecoder

